### PR TITLE
[DHT] Queries: remote values filtering

### PR DESCRIPTION
**Pull request by sim590**
This is addressing #43. The main functionnality is operationnal. I'm still working on the following:

- [x] Add documentation in the wiki about queries (how to use, use cases, benefits, ...);
- [x] Handle spaces and SELECT,WHERE lower/upper case in constructing a query from a string: `Query::Query(std::string&&)`;

### CODE REVIEW

**Comment by sim590** [comment]
- [ ] We have to be clarify this:  https://github.com/sim590/opendht/commit/f46043069d60040081bf18668d4714b19c3e2316#commitcomment-17565515

@aberaud: I think this is ready for review. When you're done, I'll squash everything in relevant seperate commits and you'll be clear to merge.

**Comment by sim590** [comment]
Travis is complaining, but it doesn't appear to be related to my changes. Or else I don't understand the error message...

File: tools/dhtscanner.cpp [hunk 1]
```diff
@@ -63,7 +63,7 @@ step(DhtRunner& dht, std::atomic_uint& done, std::shared_ptr<NodeSet> all_nodes,
         done--;
         std::cout << done.load() << " operations left, " << all_nodes->size() << " nodes found." << std::endl;
         cv.notify_one();
-    });
+    }, {}, {});
```
Comment (sim590) [inline tools/dhtscanner.cpp]: @aberaud: can you comment on this? It makes Travis happy, but the [signature](https://github.com/sim590/opendht/blob/queries/include/opendht/dhtrunner.h#L65) it's using should allow empty parameter... I don't really like this as it more looks like a work around the issue.

File: include/opendht/value.h [hunk 2]
```diff
@@ -422,6 +481,255 @@ struct Value
...
+     *  [SELECT <$field$> [WHERE <$field$=$value$>]]
+     *
+     *  where
+     *
+     *  - $field$ = *|id|value_type|owner_pk|user_type
+     *  - $value$ = $string$|$integer$
+     *  - $string$: a simple string WITHOUT SPACES.
+     *  - $integer$: a simple integer.
+     */
+    Query(std::string&& q_str);
```
Comment (aberaud) [inline include/opendht/value.h]: const std::string& q_str (value is not moved in method)

**Comment by sim590** [comment]
A result of the choice of implementation is that when creating a selection in a query, like so:
```cpp
Query q("SELECT id")
dht.get(hash, [](std::vector<std::shared_ptr<Value>>>& values) {
        /* only the field id is valid and relevant */
    }, {}, {}, std::move(q));
```
Some fields in the returned values may or may not be empty. In reality, if a value is fetched from the local node's storage, the value will have other non-empty valid fields and if another value is fetched from a remote node's storage, only the selected fields specified in the query will be valid. **This is something that has to be documented, or else a change in the API is needed if we want to provide unambiguous API**. By that I mean provide a different callback signature that would offer a consistent structure containing the results of the selection of a query...

Also, Let's say you make a `get` request with the query `SELECT id` for the target hash `h` using the `dhtnode` tool and that it happens the local node has the storage for `h`. When stacking values for the `get` callback before ef067e6, the following would happen:
```
Get: found value (after 0.000437637s)
	Value[id:7f96f114f92b4ce9 Data (type: 0 ): 66 6f 6f ]
Get: found value (after 0.00424051s)
	Value[id:7f96f114f92b4ce9 Data (type: 0 ): ]
Get: completed (took 0.00561377s)
```
The first value is from the local storage and the second is from the network (data is empty as a result of the query `SELECT id`). If OpenDHT allows empty data and that two different values from different peers can have the same value id, there is no way to differentiate the first value from the second one.

ef067e6 makes sure that no redundant information will be present, i.e. that in the example above only the first value would be present in the callback `values` argument.

**Comment by sim590** [comment]
Merged. See 51aa1de.

**Pull request closed without merging**